subsys/usb/usb_c/usbc_pe_src_states.c

```diff
@@ -0,0 +1,702 @@
+		if (policy_check_if_ps_ready(dev)) {
+			pe_set_state(dev, PE_SRC_READY);
+		}
+	}
+	/*
+	 * The Policy Engine Shall transition to the PE_SRC_Hard_Reset
+	 * state when:
+	 *	1) A Protocol Error occurs.
+	 */
+	else if (atomic_test_bit(pe->flags, PE_FLAGS_PROTOCOL_ERROR)) {
+		pe_set_state(dev, PE_SRC_HARD_RESET);
+	}
+}
+
+void pe_src_transition_supply_exit(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/* Send PS_RDY message */
+	if (pe->snk_request_reply == SNK_REQUEST_VALID) {
+		pe_send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_PS_RDY);
+		/* Explicit Contract is now in place */
+		atomic_set_bit(pe->flags, PE_FLAGS_EXPLICIT_CONTRACT);
+		/* Update present contract */
+		pe->present_contract = pe->snk_request;
+	}
+}
+
+/**
+ * @brief 8.3.3.2.6 PE_SRC_Ready State
+ */
+void pe_src_ready_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	LOG_INF("PE_SRC_Ready");
+
+	/*
+	 * If the transition into PE_SRC_Ready is the result of Protocol Error
+	 * that has not caused a Soft Reset then the notification to the
+	 * Protocol Layer of the end of the AMS Shall Not be sent since there
+	 * is a Message to be processed.
+	 *
+	 * Else on entry to the PE_SRC_Ready state the Source Shall notify the
+	 * Protocol Layer of the end of the Atomic Message Sequence (AMS).
+	 */
+	if (atomic_test_and_clear_bit(pe->flags,
+				PE_FLAGS_PROTOCOL_ERROR_NO_SOFT_RESET)) {
+		pe_dpm_end_ams(dev);
+	}
+}
+
+void pe_src_ready_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+	struct usbc_port_data *data = dev->data;
+	struct protocol_layer_rx_t *prl_rx = data->prl_rx;
+
+	/* Handle incoming messages */
+	if (atomic_test_and_clear_bit(pe->flags, PE_FLAGS_MSG_RECEIVED)) {
+		union pd_header header = prl_rx->emsg.header;
+
+		/*
+		 * Extended Message Requests
+		 */
+		if (header.extended) {
+			extended_message_not_supported(dev);
+		}
+		/*
+		 * Data Message Requests
+		 */
+		else if (header.number_of_data_objects > 0) {
+			switch (header.message_type) {
+			case PD_DATA_REQUEST:
+				pe_set_state(dev, PE_SRC_NEGOTIATE_CAPABILITY);
+				break;
+			default:
+				pe_set_state(dev, PE_SEND_NOT_SUPPORTED);
+			}
+		}
+		/*
+		 * Control Message Requests
+		 */
+		else {
+			switch (header.message_type) {
+			case PD_CTRL_GOOD_CRC:
+				/* Do nothing */
+				break;
+			case PD_CTRL_NOT_SUPPORTED:
+				/* Do nothing */
+				break;
+			case PD_CTRL_PING:
+				/* Do nothing */
+				break;
+			case PD_CTRL_GET_SOURCE_CAP:
+				pe_set_state(dev, PE_SRC_SEND_CAPABILITIES);
+				break;
+				break;
+			case PD_CTRL_DR_SWAP:
+				pe_set_state(dev, PE_DRS_EVALUATE_SWAP);
+				break;
+			/*
+			 * USB PD 3.0 6.8.1:
+			 * Receiving an unexpected message shall be responded
+			 * to with a soft reset message.
+			 */
+			case PD_CTRL_ACCEPT:
```

Comment:
Practically, I think it's OK to ignore unexpected GoodCRC and Not_Supported, but does the spec say that we should treat these differently than an unexpected Reject or Accept? Should an unexpected GoodCRC even make it past the PRL to the PE?

Reply:
Both messages are passed from the PRL to the PE and are expected in this state. Good_Crc should be ignored but it's probably a policy decision as to what to do when Not_Supported is received.